addons/udes_advanced_packaging/models/stock_picking.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+from collections import defaultdict
+
+from odoo import models, _
+from odoo.exceptions import ValidationError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _should_reserve_full_packages(self):
+        """Method to determine if picking should reserve entire packages"""
+        self.ensure_one()
+        return self.picking_type_id.u_reserve_as_packages
+
+    def _reserve_full_packages(self):
+        """
+        If the picking type of the picking in self has full package
+        reservation enabled, whole packages will be reserved.
+        """
+        Quant = self.env["stock.quant"]
+        MoveLine = self.env["stock.move.line"]
+
+        # do not reserve full packages when bypass_reserve_full packages
+        # is set in the context as True
+        if not self.env.context.get("bypass_reserve_full_packages"):
+            for picking in self.filtered(lambda p: p._should_reserve_full_packages()):
+                all_quants = Quant.browse()
+                remaining_qtys = defaultdict(int)
+
+                # get all packages
+                packages = picking.mapped("move_line_ids.package_id")
```

Comment:
Do we want to pre-cache this at the top for all relevant pickings

Reply:
Do you mean outside of the for loop? And if so, I am not sure if this would work the same way.

Reply:
picking.mapped("move_line_ids.package_id")
just this outside then its cached for all pickings then when looping through the picking it gets the relevant ones

Reply:
typo 
pickings = self.filtered(...)
pickings.mapped("move_line_ids.package_id") before
then 
for picking in pickings:
     ...
     picking.mapped("move_line_ids.package_id")
     ....